src/traces/scattergl/index.js

```diff
@@ -259,15 +259,28 @@ function sceneUpdate(gd, subplot) {
         // draw traces in proper order
         scene.draw = function draw() {
             var i;
+            var allReadyDrawLines = Lib.repeat(false, scene.count);
+            var allReadyDrawMarkers = Lib.repeat(false, scene.count);
             for(i = 0; i < scene.count; i++) {
                 if(scene.fill2d && scene.fillOptions[i]) {
                     // must do all fills first
                     scene.fill2d.draw(i);
+                    if(scene.line2d && scene.lineOptions[i] && (scene.fillOptions[i].fillmode === 'tozeroy' || scene.fillOptions[i].fillmode === 'tozerox')) {
+                        scene.line2d.draw(i);
+                        allReadyDrawLines[i] = true;
+                    }
+                    if(scene.scatter2d && scene.markerOptions[i] && (!scene.selectBatch || !scene.selectBatch[i]) && (scene.fillOptions[i].fillmode === 'tozeroy' || scene.fillOptions[i].fillmode === 'tozerox')) {
```

Comment:
Looks like this isn't 100% correct still. 

Referencing https://github.com/plotly/plotly.js/issues/2450 and codepen https://codepen.io/alexcjohnson/pen/YemVOd?editors=1010 - which doesn't render correctl.

I think our (not yours) "must do all fills first" assumption is wrong.

Reply:
i updated with my branch , and i add a new test ( https://github.com/plotly/plotly.js/pull/2979/files#diff-81d378492b6b240494eadcde98f23458 )